### PR TITLE
avoid exec_script_ahn! forking another process

### PR DESCRIPTION
**Pull request by kares**
proposal for loading script/ahn in the same process (avoiding `exec`) ... details for doing so at https://github.com/adhearsion/adhearsion/issues/532

### CODE REVIEW

**Comment by Jared-Prime** [comment]
@benlangfeld curious why some of those features have `@skip_jruby`, could you give some background on those directives?

**Comment by benlangfeld** [comment]
@Jared-Prime Those are daemonization related which isn't supported on JRuby

**Comment by Jared-Prime** [comment]
gotcha. thanks

**Comment by kares** [comment]
@benlangfeld does it make sense ... to try cleaning up those features to match the non-forked behaviour?

**Comment by benlangfeld** [comment]
I can't consider this change unless the features pass, if that's what you're asking.

**Comment by kares** [comment]
Thanks Ben, in the end it really caused regressions - now as much as I understand and specs / features cover all is the same - let me know if it's green with you and I can squash the commits with a changelog.

File: lib/adhearsion/script_ahn_loader.rb [hunk 1]
```diff
@@ -8,9 +8,22 @@ module ScriptAhnLoader
...
+        if File.exists?(script)
+          load script
+          return true
+        end
+        path = path.parent
+      end while ! path.root?
+      nil
+    end
+
     def self.exec_script_ahn!(args = ARGV)
```
Comment (benlangfeld) [inline lib/adhearsion/script_ahn_loader.rb]: It looks like nothing calls this any more. Can it be removed?

File: lib/adhearsion/cli_commands/ahn_command.rb [hunk 2]
```diff
@@ -100,28 +102,26 @@ def restart(path = nil)
...
-          args[1] = '.'
-          ScriptAhnLoader.exec_script_ahn! args
-        end
+        raise PathInvalid, path unless ScriptAhnLoader.in_ahn_application?(path)
+        # load script/ahn which than boots the Rails environment :
+        Dir.chdir(path) { ScriptAhnLoader.load_script_ahn(path) }
         path
       end
 
+      # @deprecated no longer used
```
Comment (benlangfeld) [inline lib/adhearsion/cli_commands/ahn_command.rb]: May as well just remove it.

File: bin/ahn [hunk 3]
```diff
@@ -1,4 +1,7 @@
 #!/usr/bin/env ruby
 
 require 'adhearsion/cli'
-Adhearsion::CLI::AhnCommand.start
+# when script/ahn gets loaded it does AhnCommand.start
+unless Adhearsion::ScriptAhnLoader.load_script_ahn
```
Comment (benlangfeld) [inline bin/ahn]: Maybe `Adhearsion::ScriptAhnLoader.load_script_ahn || Adhearsion::CLI::AhnCommand.start` is clearer in intent?

File: lib/adhearsion/script_ahn_loader.rb [hunk 4]
```diff
@@ -8,9 +8,22 @@ module ScriptAhnLoader
     RUBY = File.join(*RbConfig::CONFIG.values_at("bindir", "ruby_install_name")) + RbConfig::CONFIG["EXEEXT"]
     SCRIPT_AHN = File.join('script', 'ahn')
 
+    def self.load_script_ahn(path = Dir.pwd)
+      path = Pathname.new(path).expand_path
+      begin
+        script = File.join(path, SCRIPT_AHN)
```
Comment (benlangfeld) [inline lib/adhearsion/script_ahn_loader.rb]: Why not re-use `#in_ahn_application?`?

Comment (kares) [inline lib/adhearsion/script_ahn_loader.rb]: mostly cause of explicitly checking whether the script path exists before trying to load it ... seemed "less confusing" than to rely on an implementation detail of `in_ahn_application?` 

```ruby
    def self.load_script_ahn(path = Dir.pwd)
      path = Pathname.new(path).expand_path
      begin
        if in_ahn_application?(path)
          load File.join(path, SCRIPT_AHN)
          return true
        end
        path = path.parent
      end until path.root?
      nil
    end
```

... but I do not care that much, if this the preferred way for you?

Comment (benlangfeld) [inline lib/adhearsion/script_ahn_loader.rb]: It's not so much an implementation detail of that method - that method is an abstraction of what you want to do. I was just looking for ways to make this read more naturally.

Comment (kares) [inline lib/adhearsion/script_ahn_loader.rb]: OK will do as recommended if it reads better ... and squash all relevant commits with a change-log

**Comment by benlangfeld** [comment]
Looking good. I'll do a final review after a squash, changelog entry and the tests pass :)

**Comment by kares** [comment]
actually passed: https://travis-ci.org/kares/adhearsion/builds/52757410 ... but adding [proposed](https://github.com/adhearsion/adhearsion/pull/534#discussion_r25593946) `in_ahn_application?` makes it fail ;(

**Comment by benlangfeld** [comment]
No problem then @kares, leave it as it is :)

**Comment by kares** [comment]
@benlangfeld thanks for the restart ... and all the feedback!

failed again - this time on out-of-memory (seen those with JRuby before sometimes) give it another spin maybe :) ... also more JRUBY_OPTS memory tunings might help those

for the record with the proposed change the failures are: https://travis-ci.org/kares/adhearsion/jobs/52794751 - might look at those later ( but cucumber season takes me some time ) - seemed weird on first sight that it regressed

**Comment by benlangfeld** [comment]
Unfortunately this appears to have broken loading an Adhearsion application across versions, for example using the installed `ahn` at v2.6.1 to load an app bundled at Adhearsion's develop branch. This is why the forking was in place.

I'm looking at this now to see if it can be mitigated by loading less of Adhearsion at what used to be the pre-fork point, and if so I'll go ahead with that. If not, I'll have to revert this.

**Comment by benlangfeld** [comment]
See https://github.com/adhearsion/adhearsion/commit/1ce80d12fb5f79b62f45f71300b7ba7ba693487e and https://github.com/adhearsion/adhearsion/commit/66e1609eaf153d80053337c6e5235d056370a904.

If you can come up with some way of addressing your performance concerns which don't break application loading (perhaps in Adhearsion 3 where there is room to break backward compatibility), I'd love to see it, but 2.6.2 will be released shortly including this reversion.

**Comment by sfgeorge** [comment]
@benlangfeld That's unfortunate, but I understand the need to move forward with 2.6.2 for now.   Thanks for the head's up!  Hopefully we can re-integrate it in 2.x later somehow without breaking backwards compatibility.

**Comment by sfgeorge** [comment]
> Unfortunately this appears to have broken loading an Adhearsion application across versions, for example using the installed `ahn` at v2.6.1 to load an app bundled at Adhearsion's develop branch. This is why the forking was in place.

@benlangfeld Does this relate to `ahn create` or `ahn start`?  Can you give us an example of this use case?

**Comment by benlangfeld** [comment]
It would be `ahn start`, but binstubs obviate this issue.

**Comment by sfgeorge** [comment]
Phew, that's great news...  I was hoping that binstubs would resolve the issue.

In that case, @kares and I would like to discuss re-introducing the no-spawn optimization to ahn 3.0.0.RC.  Here are some important reasons for us to try to reincorporate the solution that prevents forking.  Forking:
1. Slows down booting
2. Makes it more difficult to get the correct adhearsion pid to monitor.
3. Under JRuby, duplicates certain debugging instrumentation in both processes, causing adhearsion it to crash.  As a result, you simply can’t use certain standard JVM debugging tools on Adhearsion that you otherwise could.

**Comment by kares** [comment]
+1 would be great to have this reconsidered.

its really bad for running AHN under JRuby with this in place (on a remote host).
setting a JAVA/JRUBY_OPTS prior to process starting gets quite complicated  :
- one has to account for this with memory settings (not to set initial heap `-Xms` too high)
- gets a 'dirty' gc.log when trying to `-J-XX:+PrintGCDetails` to a file (from two processes)
- same with any JRuby runtime logging e.g. `-Xlog.exceptions=true` - gets noisy from starter process
- can not use any (JMX) remote monitoring easily as ports from the first JVM would be occupied
  need to know port numbers ahead-of-time e.g. `-Dcom.sun.management.jmxremote.port=7091`

**Pull request merged**